mackerel-plugin-multicore/multicore.go

```diff
@@ -0,0 +1,402 @@
+
+	if lastValue <= value {
+		return diff, nil
+	}
+	return 0.0, nil
+}
+
+func fetchLoadavg5() (float64, error) {
+	contentbytes, err := ioutil.ReadFile("/proc/loadavg")
+	if err != nil {
+		return 0.0, err
+	}
+	content := string(contentbytes)
+	cols := strings.Split(content, " ")
+
+	f, err := strconv.ParseFloat(cols[1], 64)
+	if err != nil {
+		return 0.0, err
+	}
+
+	return f, nil
+}
+
+func printValue(key string, value float64, time time.Time) {
+	fmt.Printf("%s\t%f\t%d\n", key, value, time.Unix())
+}
+
+func outputCPUUsage(cpuUsage map[string]*cpuPercentages, now time.Time) {
+	if cpuUsage != nil {
+		for key, values := range cpuUsage {
+			if key != "cpu" {
+				printValue(fmt.Sprintf("multicore.cpu.%s.user", key), values.User, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.nice", key), values.Nice, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.system", key), values.System, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.idle", key), values.Idle, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.iowait", key), values.IoWait, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.irq", key), values.Irq, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.softirq", key), values.SoftIrq, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.steal", key), values.Steal, now)
+				printValue(fmt.Sprintf("multicore.cpu.%s.guest", key), values.Guest, now)
+			}
+		}
+	}
+}
+
+func outputAverageCPUUsage(cpuUsage map[string]*cpuPercentages, now time.Time) {
+	if cpuUsage != nil {
+		values := cpuUsage["cpu"]
+		printValue("multicore.average_per_core.user", values.User, now)
+		printValue("multicore.average_per_core.nice", values.Nice, now)
+		printValue("multicore.average_per_core.system", values.System, now)
+		printValue("multicore.average_per_core.idle", values.Idle, now)
+		printValue("multicore.average_per_core.iowait", values.IoWait, now)
+		printValue("multicore.average_per_core.irq", values.Irq, now)
+		printValue("multicore.average_per_core.softirq", values.SoftIrq, now)
+		printValue("multicore.average_per_core.steal", values.Steal, now)
+		printValue("multicore.average_per_core.guest", values.Guest, now)
+	}
+}
+
+func outputLoadavgPerCore(loadavgPerCore float64, now time.Time) {
+	printValue("multicore.loadavg_per_core.loadavg5", loadavgPerCore, now)
+}
+
+func outputDefinitions() {
+	fmt.Println("# mackerel-agent-plugin")
+	var graphs mp.GraphDef
+	graphs.Graphs = graphDef
+
+	b, err := json.Marshal(graphs)
+	if err != nil {
+		log.Fatalln("OutputDefinitions: ", err)
+	}
+	fmt.Println(string(b))
+}
+
+// main function
+func main() {
+	optTempfile := flag.String("tempfile", "", "Temp file name")
+	var tempFileName string
+	if *optTempfile != "" {
+		tempFileName = *optTempfile
+	} else {
+		tempFileName = "/tmp/mackerel-plugin-multicore"
+	}
+	now := time.Now()
+
+	currentValues, _ := collectProcStatValues()
+	lastValues, lastTime, err := fetchLastValues(tempFileName)
+	saveValues(tempFileName, currentValues, now)
```

Comment:
When `collectProcStatValues` fails, this line saves `nil` to the temporary file.